Scale Inspector

NanaKwame opened the pull request:
Goal is to design the scale inspector.


Current state:
- first pass at showing just the fields we have access to

NanaKwame commented:
As of now only the first 4 fields show something. Need to figure out how to get more fields to get populated.

ktbacher commented on src/js/components/inspectors/Scale.tsx:
```diff
@@ -20,22 +21,41 @@ interface StateProps {
             <li>type: {scale.get('type')}</li>
-            {/* <li>range: {scale.get('range')}</li> TODO */}
+
+            <Property name='type' label='Type' type='select' opts={typeOpts} {...props} />
```
I'm not sure how the onChange works here, doesn't seem to be updating the property

ktbacher commented on src/js/components/inspectors/FormInputProperty.tsx:
```diff
@@ -189,6 +189,9 @@ class BaseFormInputProperty extends React.Component<OwnProps & StateProps & Disp
         );
 
       case 'select':
+        if (value.scheme) {
```
Might be a little hacky, open to suggestions on other ways to handle this.  `value = {scheme: "someColorScheme"}` otherwise and threw off the selector options

ktbacher commented on src/js/components/inspectors/Scale.tsx:
```diff
@@ -18,29 +21,103 @@ interface StateProps {
+    this.props.updateScaleProperty(scaleId, property, value);
+  };
+
+  public handleRangeColor(evt) {
```
not sure if there is a way to handle this in the other `handleChange()` function but need to pass an object for the color scheme value

NanaKwame replied:
gotcha. moved it to handlechange. lmk if it looks ok.

jonathanzong commented on src/js/store/factory/Scale.ts:
```diff
@@ -16,7 +16,8 @@ export type LyraScale = {
   _domain: any[],
-  _range: any[]
+  _range: any[],
+  nice: any
```
as far as i know, no need to add this here. it's defined in the definitions of e.g. `ThresholdScale`, `QuantizeScale`, etc, which are part of `RangeScale`, and included when we use `& RangeScale` to sort of extend `RangeScale`.

jonathanzong commented on src/js/reducers/scalesReducer.ts:
```diff
@@ -31,7 +31,13 @@ export function scalesReducer(state: ScaleState, action: ActionType<typeof scale
+    if (state.getIn([String(id), ...p.property.split(".")])) {
+      return state.setIn([String(id), ...p.property.split(".")], p.value);
+    }
+    const key = p.property.split(".").slice(-1).toString();
+    const obj = {};
+    obj[key] = p.value;
+    return state.setIn([String(id), ...p.property.split(".").slice(0,-1)], obj);
```
these lines break things right now. to remind me, can you write down the cases that you're trying to handle here?

ktbacher replied:
Yeah this is what I needed some help with... Trying to handle the `range = {scheme: someColorScheme}` vs `range = someString`

jonathanzong replied:
cool -- i think you can change this from an if statement to a try catch to handle the `TypeError`:
```
try {
      return state.setIn([String(id), ...p.property.split(".")], p.value);
    }
    catch (e) {
      const key = p.property.split(".").slice(-1).toString();
      const obj = {};
      obj[key] = p.value;
      return state.setIn([String(id), ...p.property.split(".").slice(0,-1)], obj);
    }
```

jonathanzong replied:
and maybe leave a comment quickly explaining the typescript issue and reference the range and range.scheme example

NanaKwame commented:
In my recent commit:
- added keypress event support to FormInput, Property and MultiValueProperty.
- added onblur support to Property, MultiValueProperty
- added code export.ts to export the right vega specs for manual vs data field domain

NanaKwame commented:
@jonathanzong  Currently the questions I have are:
- Do I need to fix MultiValueProperty.test.jsx? If so how do I run this to make sure it is doing the right thing?
- I went a bit overboard with the event support but I think both onBlur and onKeyPress are valuable. lmk?

@ktbacher do we still need the value.scheme fix. I am getting an error because value is declared as a const. I might also have missed some updates with my merge conflict resolution.

NanaKwame commented on src/js/components/inspectors/MultiValueProperty.test.jsx:
```diff
@@ -0,0 +1,239 @@
+/* eslint no-unused-expressions:0,new-cap:0 */
```
not sure if I should get these tests to work or how do I run them?

jonathanzong replied:
the tests are stale so ignore them!

NanaKwame commented on src/js/store/factory/Scale.ts:
```diff
@@ -16,7 +16,9 @@ export type LyraScale = {
-  _range: any[]
+  _range: any[],
+  _manual: boolean,
+  _domainArray: any[]
```
introduced this so we don't need to override _domain is the user switches to manual. That way if they toggle the manual check the _domain value with fields isn't lost. But we resolve this in the exports.ts file.

jonathanzong replied:
is `_manualDomain` a clearer variable name for `_domainArray`?

NanaKwame commented on src/js/components/inspectors/MultiValueProperty.tsx:
```diff
@@ -0,0 +1,178 @@
+    let arrayValues = [...props.value];
+    arrayValues.push(newVal);
+
+    evt.target.name = props.name;
```
working on a better way to go about this.

jonathanzong commented on src/js/components/inspectors/FormInputProperty.tsx:
```diff
@@ -21,6 +21,7 @@ interface OwnProps {
   signal?: string;
   onChange?: (e) => any; // TODO: find function in/out types
   onBlur?: () => any; // TODO: find function in/out types
+  onKeyPress?: () => any; // TODO: find function in/out types
```
these should all be `=> void` if you wanna resolve the TODOs :)

jonathanzong commented on src/js/components/inspectors/Property.tsx:
```diff
@@ -71,7 +73,7 @@ function mapStateToProps(reduxState: State, ownProps: OwnProps): StateProps {
   const scaleName = scale && getInVis(reduxState, 'scales.' + scale + '.name');
 
-  const value = getIn(state, path);
+  const value = ownProps.value === undefined ? getIn(state, path) : null;
```
if you read this and then read line 79, this change appears to amount to this effect: return ownProps.value if it's not undefined, else get it from the store. if the store value is null, return undefined. If that's what we want, we should try rewriting it as a less complex expression

NanaKwame replied:
oh yeah i missed that. thnx!

jonathanzong commented on src/js/components/inspectors/Scale.tsx:
```diff
@@ -1,46 +1,173 @@
   primId: number;
   primType: PrimType;
-
+  _manual: boolean;
```
no references to this -- remove?

jonathanzong commented on src/js/components/inspectors/Scale.tsx:
```diff
@@ -1,46 +1,173 @@
 interface StateProps {
   scale: ScaleRecord;
+  fields: any;
+  fieldMap: any;
```
fields: string[];
fieldMap: Map<string, string[]>;

jonathanzong commented on src/js/components/inspectors/MultiValueProperty.tsx:
```diff
@@ -0,0 +1,178 @@
+          onDragEnter={() => this.setState({ dragOver: ++dragOver })}
+          onDragLeave={() => this.setState({ dragOver: --dragOver })}>
+          {selectedValsEl}
+          <Property {...props} label="Add" value={null} onChange={props.onChange ? (e) => this.handleChange(e) : null} onBlur={props.onBlur ? (e) => this.handleChange(e) : null} onKeyPress={props.onKeyPress ? (e) => this.handleChange(e) : null} />
```
why `onKeyPress`? did the existing `onChange` handling in `FormInputProperty` not work?

NanaKwame replied:
so the issue with onchange is that it fires for every change(even while you are typing) which we don't want for the manual text input. so just adding support for keypress so that we can listen for the enter key.

jonathanzong commented on src/js/components/inspectors/MultiValueProperty.test.jsx:
```diff
@@ -0,0 +1,239 @@
+// /* eslint no-unused-expressions:0,new-cap:0 */
```
i just realized that i misunderstood your question earlier. i thought you were asking about old tests, but i see that you wrote new tests!

i've never run the tests before but i think that you can try to run these using `yarn test`. see `package.json` for a list of scripts that you can run with yarn

NanaKwame replied:
so i didn't actually write this. i just copied them from property and did a string search replace so they are actually wrong i think. :/ so will probably leave it commented out.

NanaKwame replied:
wdyt?

jonathanzong replied:
ah ok, that's fine to just leave it then! all the tests haven't been updated since before the refactor, and since we're actively making changes at a fast pace, it's not worth investing much time into updating the tests until things are a bit more stable IMO

jonathanzong commented on src/js/components/inspectors/MultiValueProperty.tsx:
```diff
@@ -0,0 +1,178 @@
+// hunch: probably actually needs to be a bunch of different types anded to a base type
+// instead of one type with everything optional
+// but it would take a long time to figure out what the mutually exclusive fields are
+interface OwnProps {
```
this props interface is a mess because `Property.tsx` doesn't have a consistent interface across all its uses. Since we're writing this new component, and can control where it's used, let's clean up the props interface to remove optionals and remove props that aren't used here.

jonathanzong commented on src/js/components/inspectors/MultiValueProperty.tsx:
```diff
@@ -0,0 +1,178 @@
+  scale?: any;
+  srcField?: any;
+  scaleName?: any;
+  dragging?: DraggingStateRecord;
```
For example, do you want people to be able to drag fields or signals onto the `MultiValueProperty`? If not, let's not leave this prop here.

NanaKwame replied:
Gotcha. So I am going to take this as strip it down to just what we need for now and we can add more stuff later if we need it? Gonna go with that for now.

NanaKwame replied:
I removed a bunch of stuff. Lmk if I should remove more or put some stuff back.

NanaKwame commented:
known console errors:
- `value` prop on `input` should not be null...
- A component is changing a controlled input of type checkbox to be uncontrolled...

looking into these but I think we can merge with these. could be separate issues.

jonathanzong commented on src/js/components/inspectors/Property.tsx:
```diff
@@ -71,12 +73,10 @@ function mapStateToProps(reduxState: State, ownProps: OwnProps): StateProps {
     group:  getIn(state, path + '.group'),
     signal: getIn(state, path + '.signal'),
-    value:  value !== null && value !== undefined ? value : ownProps.value,
+    value: ownProps.value === undefined ? getIn(state, path) : undefined, // Not using || because value could be 0
```
observation: it's now no longer possible for the value of `ownProps.value` to be directly returned if it's not `undefined`. is that what we want?

NanaKwame replied:
thnx for putting in the fix for this.
